src/gnu/prolog/database/Module.java

```diff
@@ -44,6 +55,86 @@
 	protected List<Pair<PrologTextLoaderError, Term>> initialization = Collections
 			.synchronizedList(new ArrayList<Pair<PrologTextLoaderError, Term>>());
 
+	protected AtomTerm name = null;
+	protected List<CompoundTermTag> exports = null;
+	public final static AtomTerm userAtom = AtomTerm.get("user");
+
+	/** PredicateTag to code mapping */
+	protected Map<CompoundTermTag, PrologCode> tag2code = new HashMap<CompoundTermTag, PrologCode>();
+	protected final Map<CompoundTermTag, List<PrologCodeListenerRef>> tag2listeners = new HashMap<CompoundTermTag, List<PrologCodeListenerRef>>();
+
+
+	public Module(AtomTerm name, List<CompoundTermTag> exports)
+	{
+		this.name = name;
+		this.exports = exports;
+	}
+
+	public void importPredicates(Environment environment, AtomTerm exportingModule, List<CompoundTermTag> exports) throws PrologException
+	{
+		environment.pushModule(name);
+		try
+		{
+			for (CompoundTermTag export: exports)
+			{
+				importPredicate(environment, exportingModule, export);
+			}
+		}
+		catch(Throwable builtin)
+		{
+			// FIXME: It is quite bad if this happens
```

Comment:
Ah, quite. Digging deeper into this uncovered some other problems that I meant to get to the bottom of but never did. I think it's much clearer now, and with the added bonus of that exception not disappearing into the ether. Have a look at the revised version with my extra commit.